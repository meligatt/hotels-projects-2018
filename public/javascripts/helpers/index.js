import axios from 'axios';

// TODO: maybe add an array as argument instead of comparison function
// check if is working for strings
export const sortAscendent = ({array, orderBy}) => {
  const sortedAarray = array.sort((a,b) => {
    // previous implementation doesnt work for strings:
    // return a[orderBy] - b[orderBy];
    if(a[orderBy] < b[orderBy]) return -1;
    if(a[orderBy] > b[orderBy]) return 1;
    return 0;
  });
  return sortedAarray;
}

export const sortDescendent = ({array, orderBy}) => {
  const sortedAarray = array.sort((a,b) => {
    // previous implementation doesnt work for strings:
    // return b[orderBy] - a[orderBy];

    if(a[orderBy] > b[orderBy]) return -1;
    if(a[orderBy] < b[orderBy]) return 1;
    return 0;
  })
  return sortedAarray;
}

export function makeRequest({ endpoint, method, params }){
  // TODO: add this hardcoded value in a config file with URL's
  const APP_URL = "http://localhost:3000";

 return axios({
   url: APP_URL + endpoint,
   method: method,
   params,
   transformResponse: [(data) => data],
 })
 .then(({data}) => {
   return JSON.parse(data);
 })
 .catch((error) => {
   return error;
 });
}
