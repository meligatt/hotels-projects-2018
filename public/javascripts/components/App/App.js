import axios from 'axios';
import React, { Component } from 'react';
import FilterBar from '../FilterBar';
import HotelList from '../HotelList';
import { sortDescendent, sortAscendent, makeRequest } from '../../helpers';

class App extends Component {
  constructor(props){
    super(props);
    this.state = {
      hotels: null,
    }
  }

  componentDidMount(){
     makeRequest({
       endpoint:'/api/results',
       method:"GET"
     }).then((data) => {
       this.setState({
       hotels: sortDescendent({array: data, orderBy: "price"}),
       optionSelected: "desc",
       dataReady: true,
     });
   }).catch((error) => { console.log(error)})
  }

   handleOptionChange(sortValue){
    const { hotels } =  this.state;

    if (sortValue === "asc") {
      return this.setState({
        hotels: sortAscendent({array: hotels, orderBy: "price"}),
        optionSelected: sortValue,
      });
    }

    this.setState({
      hotels: sortDescendent({array: hotels, orderBy: "price"}),
      optionSelected: sortValue,
    });
  }

  render(){
    const {hotels, optionSelected } = this.state;
    if(hotels === null || hotels.length === 0) {
      return(<div>no hotels found</div>)
    }
    return(
      <div>
      <header>
        <h1>Qantas Hotels</h1>
      </header>
      <main>
      <FilterBar
        optionSelected={ optionSelected }
        onOptionChange = { (event) => { this.handleOptionChange(event.target.value) } }
      />
      <HotelList hotels={ hotels }/>
      </main>
      </div>
    )
  }
}
export default App;
