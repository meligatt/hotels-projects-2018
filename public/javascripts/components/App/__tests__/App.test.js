import React from 'react';
import App from '../App';

//TODO Move this imports to setup file
import axios from 'axios';
import httpAdapter from 'axios/lib/adapters/http';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { sortDescendent, makeRequest } from '../../../helpers';

// Mocking the imported module:
jest.mock('../../../helpers', () => (
  { makeRequest: jest.fn(),
    sortDescendent: jest.fn(),
   })
 )

//TODO refactor this set up
configure({ adapter: new Adapter() });
axios.defaults.host = 'http://localhost:3000';
axios.defaults.adapter = httpAdapter;


describe("App", () => {
  describe("when state is initial", () => {
    it("renders no items in the list", () => {
      const mockData = [];
      const promise = Promise.resolve(mockData);
      makeRequest.mockImplementation(() => promise);
      sortDescendent.mockImplementation(() => mockData);

      const wrapper = shallow(<App />);

      return promise.then((data) => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe("when state is populated with data", () => {
    it("renders correctly with items", () => {
      expect.assertions(2);
      const mockData = [{
        "id": "ID",
        "title": "TITLE",
        "address": "ADDRESS",
        "image": "IMG_URL",
        "rating": 4,
        "ratingType": "RATNG_TYPE",
        "promotion": "PROMOTION",
        "name": "NAME",
        "price": 329,
        "savings": 30,
        "freeCancellation": true
      },
      {
        "id": "ID2",
        "title": "TITLE2",
        "address": "ADDRESS2",
        "image": "IMG_URL2",
        "rating": 4,
        "ratingType": "RATNG_TYPE",
        "promotion": "PROMOTION",
        "name": "NAME",
        "price": 329,
        "savings": 30,
        "freeCancellation": true
      }];

      const promise = Promise.resolve(mockData);

      makeRequest.mockImplementation(() => promise);
      sortDescendent.mockImplementation(() => mockData);

      const wrapper = shallow(<App />);

      return promise.then((data) => {
        expect(wrapper.state('hotels').length).toBe(2);
        expect(wrapper).toMatchSnapshot();
      });
    })
  })
})
