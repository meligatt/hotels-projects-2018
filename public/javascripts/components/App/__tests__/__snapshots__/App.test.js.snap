// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App when state is initial renders no items in the list 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "no hotels found",
    },
    "ref": null,
    "rendered": "no hotels found",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "no hotels found",
      },
      "ref": null,
      "rendered": "no hotels found",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`App when state is populated with data renders correctly with items 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header>
          <h1>
            Qantas Hotels
          </h1>
        </header>,
        <main>
          <FilterBar
            onOptionChange={[Function]}
            optionSelected="desc"
          />
          <HotelList
            hotels={
              Array [
                Object {
                  "address": "ADDRESS",
                  "freeCancellation": true,
                  "id": "ID",
                  "image": "IMG_URL",
                  "name": "NAME",
                  "price": 329,
                  "promotion": "PROMOTION",
                  "rating": 4,
                  "ratingType": "RATNG_TYPE",
                  "savings": 30,
                  "title": "TITLE",
                },
                Object {
                  "address": "ADDRESS2",
                  "freeCancellation": true,
                  "id": "ID2",
                  "image": "IMG_URL2",
                  "name": "NAME",
                  "price": 329,
                  "promotion": "PROMOTION",
                  "rating": 4,
                  "ratingType": "RATNG_TYPE",
                  "savings": 30,
                  "title": "TITLE2",
                },
              ]
            }
          />
        </main>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h1>
            Qantas Hotels
          </h1>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Qantas Hotels",
          },
          "ref": null,
          "rendered": "Qantas Hotels",
          "type": "h1",
        },
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <FilterBar
              onOptionChange={[Function]}
              optionSelected="desc"
            />,
            <HotelList
              hotels={
                Array [
                  Object {
                    "address": "ADDRESS",
                    "freeCancellation": true,
                    "id": "ID",
                    "image": "IMG_URL",
                    "name": "NAME",
                    "price": 329,
                    "promotion": "PROMOTION",
                    "rating": 4,
                    "ratingType": "RATNG_TYPE",
                    "savings": 30,
                    "title": "TITLE",
                  },
                  Object {
                    "address": "ADDRESS2",
                    "freeCancellation": true,
                    "id": "ID2",
                    "image": "IMG_URL2",
                    "name": "NAME",
                    "price": 329,
                    "promotion": "PROMOTION",
                    "rating": 4,
                    "ratingType": "RATNG_TYPE",
                    "savings": 30,
                    "title": "TITLE2",
                  },
                ]
              }
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "onOptionChange": [Function],
              "optionSelected": "desc",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "hotels": Array [
                Object {
                  "address": "ADDRESS",
                  "freeCancellation": true,
                  "id": "ID",
                  "image": "IMG_URL",
                  "name": "NAME",
                  "price": 329,
                  "promotion": "PROMOTION",
                  "rating": 4,
                  "ratingType": "RATNG_TYPE",
                  "savings": 30,
                  "title": "TITLE",
                },
                Object {
                  "address": "ADDRESS2",
                  "freeCancellation": true,
                  "id": "ID2",
                  "image": "IMG_URL2",
                  "name": "NAME",
                  "price": 329,
                  "promotion": "PROMOTION",
                  "rating": 4,
                  "ratingType": "RATNG_TYPE",
                  "savings": 30,
                  "title": "TITLE2",
                },
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "main",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header>
            <h1>
              Qantas Hotels
            </h1>
          </header>,
          <main>
            <FilterBar
              onOptionChange={[Function]}
              optionSelected="desc"
            />
            <HotelList
              hotels={
                Array [
                  Object {
                    "address": "ADDRESS",
                    "freeCancellation": true,
                    "id": "ID",
                    "image": "IMG_URL",
                    "name": "NAME",
                    "price": 329,
                    "promotion": "PROMOTION",
                    "rating": 4,
                    "ratingType": "RATNG_TYPE",
                    "savings": 30,
                    "title": "TITLE",
                  },
                  Object {
                    "address": "ADDRESS2",
                    "freeCancellation": true,
                    "id": "ID2",
                    "image": "IMG_URL2",
                    "name": "NAME",
                    "price": 329,
                    "promotion": "PROMOTION",
                    "rating": 4,
                    "ratingType": "RATNG_TYPE",
                    "savings": 30,
                    "title": "TITLE2",
                  },
                ]
              }
            />
          </main>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h1>
              Qantas Hotels
            </h1>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Qantas Hotels",
            },
            "ref": null,
            "rendered": "Qantas Hotels",
            "type": "h1",
          },
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <FilterBar
                onOptionChange={[Function]}
                optionSelected="desc"
              />,
              <HotelList
                hotels={
                  Array [
                    Object {
                      "address": "ADDRESS",
                      "freeCancellation": true,
                      "id": "ID",
                      "image": "IMG_URL",
                      "name": "NAME",
                      "price": 329,
                      "promotion": "PROMOTION",
                      "rating": 4,
                      "ratingType": "RATNG_TYPE",
                      "savings": 30,
                      "title": "TITLE",
                    },
                    Object {
                      "address": "ADDRESS2",
                      "freeCancellation": true,
                      "id": "ID2",
                      "image": "IMG_URL2",
                      "name": "NAME",
                      "price": 329,
                      "promotion": "PROMOTION",
                      "rating": 4,
                      "ratingType": "RATNG_TYPE",
                      "savings": 30,
                      "title": "TITLE2",
                    },
                  ]
                }
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "onOptionChange": [Function],
                "optionSelected": "desc",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "hotels": Array [
                  Object {
                    "address": "ADDRESS",
                    "freeCancellation": true,
                    "id": "ID",
                    "image": "IMG_URL",
                    "name": "NAME",
                    "price": 329,
                    "promotion": "PROMOTION",
                    "rating": 4,
                    "ratingType": "RATNG_TYPE",
                    "savings": 30,
                    "title": "TITLE",
                  },
                  Object {
                    "address": "ADDRESS2",
                    "freeCancellation": true,
                    "id": "ID2",
                    "image": "IMG_URL2",
                    "name": "NAME",
                    "price": 329,
                    "promotion": "PROMOTION",
                    "rating": 4,
                    "ratingType": "RATNG_TYPE",
                    "savings": 30,
                    "title": "TITLE2",
                  },
                ],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "main",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
